text/3493-precise-pre-release-cargo-update.md

```diff
@@ -0,0 +1,146 @@
+
+| Cargo.toml spec | Cargo.lock version | Target version | Selected by cargo update  | Selected by cargo update --precise  |
+| --------------- | ------------------ | -------------- | ------------------------- | ----------------------------------- |
+| `^a.b.c`         | `a.b.c`            | `x.y.z`        | ✅                        | ✅                                  |
+| `^a.b.c`         | `a.b.c`            | `x.y.z-pre.0`   | ❌                        | ✅                                  |
+| `^a.b.c`         | `x.y.z-pre.0`       | `x.y.z-pre.1`   | ❌                        | ✅                                  |
+| `^a.b.c-pre.0`    | `a.b.c-pre.0`       | `a.b.c-pre.1`   | ✅¹                       | ✅                                  |
+| `^a.b.c-pre.0`    | `a.b.c-pre.0`       | `x.y.z`        | ✅¹                       | ✅                                  |
+| `^a.b.c`         | `a.b.c`            | `a.b.c-pre.0`   | ❌                        | ❌                                  |
+
+✅: Will upgrade
+
+❌: Will not upgrade
+
+¹This behaviour is considered by some to be undesirable and may change as proposed in [RFC: Precise Pre-release Deps](https://github.com/rust-lang/rfcs/pull/3263).
+This RFC preserves this behaviour to remain backwards compatible.
+Since this RFC is concerned with the behaviour of `cargo update --precise` changes to bare `cargo update` made in future RFCs should have no impact on this proposal.
+
+To determine if a version can be selected with `--precise` for a specification that isn't listed above cosider where pre-releases exist within version ranges.
+
+For example consider the version `~1.2.3`.
+The range for `~1.2.3` is [stated in the cargo book](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#tilde-requirements).
+
+```
+~1.2.3  := >=1.2.3, <1.3.0
+```
+
+Intuitively `1.2.4-pre.0` satisfies this inequality, therefore it can be selected with `cargo update --precise`.
+Since it is a pre-release and the specification is not, `1.2.4-pre.0` would not be selected by a bare `cargo update`.
+`1.3.0-pre.0` also satisfies the inequality but `1.2.3-pre.0` and `1.3.1-pre.0` do not.
+
+Put in simple terms the relationship between a pre-release and its stable release is always `a.b.c-pre.0 < a.b.c`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Pre-release versions are not easily auditable when they are only specified in the lock file.
+  A change that makes use of a pre-release version may not be noticed during code review as reviewers don't always check for changes in the lock file.
+- Library crates that require a pre-release version are not well supported since their lock files are ignored by their users (see [future-possibilities])
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The main alternative to this would be to accept that pre-release versions are not very usable and discourage their use.
```

Comment:
This isn't quite my take.  My concern is that we need to look at the wider needs of pre-release to make sure we are providing steps towards a coherent solution.  Personally, I would like to see more requirements gathering that leads to such a rough sketch before we move forward with changing pre-releae behavior so we don't do patch upon patch and trap ourselves in a corner.

Reply:
That makes sense. I've removed that line https://github.com/rust-lang/rfcs/pull/3493/commits/b2450dc3684792b58ec3a10a78cdfaa9d9305ba6

Reply:
While that fixes the wording, my concern still stands about whether we have enough context for moving forward with this RFC.

Reply:
As in the motivation concerning transitive dependencies not being compelling enough? How would you envisage getting this context? I'd be happy to collaborate on that effort.

Reply:
Its not whether the problem is compelling but where are we heading and is this solution aligned with that or going to cause us problems in the future?

Reply:
For myself, I would like to see that investigation done before moving forward with this RFC.  That is likely to happen through a variety of mediums.  Ideally, that would be done even before the RFC or as part of the Pre-RFC.